src/Cryptol/Symbolic/SBV.hs

```diff
@@ -114,135 +118,217 @@ proverError :: String -> M.ModuleCmd (Maybe String, ProverResult)
+                     SatQuery AllSat -> Nothing
+                     ProveQuery -> Nothing
+
+     provers <-
+       case pcProverName of
+         "any" -> M.io SBV.sbvAvailableSolvers
+         "sbv-any" -> M.io SBV.sbvAvailableSolvers
+         _ -> return [(lookupProver pcProverName)
+                        { SBV.transcript = pcSmtFile
+                        , SBV.allSatMaxModelCount = mSatNum
+                        }]
+
+     let provers' = [ p { SBV.timing = SaveTiming pcProverStats
+                        , SBV.verbose = pcVerbose
+                        , SBV.validateModel = pcValidate
+                        }
+                    | p <- provers
+                    ]
+
+     case pcQueryType of
+        ProveQuery -> runMultiProvers pc provers' SBV.proveWithAny thmSMTResults x
+        SatQuery sn -> case sn of
+          SomeSat 1 -> runMultiProvers pc provers' SBV.satWithAny satSMTResults x
+          _         -> runSingleProver pc provers' SBV.allSatWith allSatSMTResults x
+
+
+-- | Prepare a symbolic query by symbolically simulating the expression found in
+--   the @ProverQuery@.  The result will either be an error or a list of the types
+--   of the symbolic inputs and the symbolic value to supply to the solver.
+--
+--   Note that the difference between sat and prove queries is reflected later
+--   in `runProver` where we call different SBV methods depending on the mode,
+--   so we do _not_ negate the goal here.  Moreover, assumptions are added
+--   using conjunction for sat queries and implication for prove queries.
+--
+--   For safety properties, we want to add them as an additional goal
+--   when we do prove queries, and an additional assumption when we do
+--   sat queries.  In both cases, the safety property is combined with
+--   the main goal via a conjunction.
+prepareQuery ::
+  Eval.EvalOpts ->
+  M.ModuleEnv ->
+  ProverCommand ->
+  IO (Either String ([FinType], SBV.Symbolic SBV.SVal))
+prepareQuery evo modEnv ProverCommand{..} =
+  do let extDgs = M.allDeclGroups modEnv ++ pcExtraDecls
+     let tyFn = case pcQueryType of
+           SatQuery _ -> existsFinType
+           ProveQuery -> forallFinType
+     let addAsm = case pcQueryType of
+           ProveQuery -> \x y -> SBV.svOr (SBV.svNot x) y
+           SatQuery _ -> \x y -> SBV.svAnd x y
+     let ?evalPrim = evalPrim
+     case predArgTypes pcSchema of
+       Left msg -> return (Left msg)
+       Right ts ->
+         do when pcVerbose $ logPutStrLn (Eval.evalLogger evo) "Simulating..."
+            pure $ Right $ (ts,
+              do -- Compute the symbolic inputs, and any domain constraints needed
+                 -- according to their types.
+                 (args, asms) <- runWriterT (mapM tyFn ts)
+                 -- Run the main symbolic computation.  First we populate the
+                 -- evaluation environment, then we compute the value, finally
+                 -- we apply it to the symbolic inputs.
```

Comment:
I expected this comment to say something explicitly about what we do with the safety predicate. I had to think a bit to convince myself that, yes, `svAnd safety b` is actually what you want for both `:prove` and `:sat`. It wouldn't hurt to say that explicitly.

Reply:
The haddock comment for the function says pretty much that.

Reply:
I totally missed that. I guess I need a taller screen.

Reply:
I think most of my confusion was due to forgetting that "assumptions" does not mean "safety predicates". It took some effort to remember that "assumptions" come from `Z n` types.